dragnet/_readability.cc

```diff
@@ -0,0 +1,128 @@
+
+#include <vector>
+#include <string>
+#include <unordered_map>
+
+// NOTE: features is the return vector.  we assume it is preinitialized
+// to length(nblocks) with 0.0
+
+void _readability_features(
+    std::vector<uint32_t>& block_text_len,
+    std::vector<std::vector<std::pair<uint32_t, int> > >& block_readability_class_weights,
+    std::vector<std::vector<uint32_t> >& block_ancestors,
+    std::vector<std::string>& block_start_tag,
+    std::vector<double>& block_link_density,
+    int& nblocks,
+    double* features)
+{
+    std::unordered_map<uint32_t, double> scores;
+    // link density in subtrees.
+    // first entry is total text length * link density,
+    // second is total text length
+    std::unordered_map<uint32_t, std::pair<double, double> > ld;
```

Comment:
structs are annoying in cython and since this is called from cython I'm going to leave as a pair (structs need their own cdef, etc, whereas pairs get automatically converted to tuples).